### PR TITLE
Netsim: Handle Lost Connection

### PR DESCRIPTION
**Pull request by islemaster**
I did a version of this very early, and recently realized that it's not working.  So instead of using an observer and some fancy reconnect logic, I'm just wiring up an `onLostConnection` callback.  The callback gets fired whenever a node's heartbeat update fails, or self-update fails.

Not critical to see this go in for tomorrow's demo, but I figured I'd try.

### CODE REVIEW

File: apps/src/netsim/NetSimEntity.js [hunk 1]
```diff
@@ -38,13 +36,6 @@ var NetSimEntity = module.exports = function (shard, entityRow) {
    * @type {number}
    */
   this.entityID = entityRow.id;
-
-  /**
-   * Change event fired when entity's state changes in a way that
-   * should be reported.
-   * @type {ObservableEvent}
-   */
-  this.onChange = new ObservableEvent();
```
Comment (islemaster) [inline apps/src/netsim/NetSimEntity.js]: This change removes the only consumer of `NetSimEntity.onChange` so I figured it was time to tear it out.

File: apps/src/netsim/NetSimLobby.html.ejs [hunk 2]
```diff
@@ -7,6 +7,7 @@
   <div id="shard_view">
     <label for="netsim_shard_select">My Section:</label>
     <select id="netsim_shard_select"></select>
+    <span id="netsim_not_connected_note">You are not connected. Select a class section from the list to connect.</span>
```
Comment (islemaster) [inline apps/src/netsim/NetSimLobby.html.ejs]: Yes, this is pretty lame; there's already some UI pain around the lobby component, I think I'd like to wait until I break it apart to do something different with this "disconnected" message.

File: apps/src/netsim/NetSimConnection.js [hunk 3]
```diff
@@ -201,6 +201,16 @@ NetSimConnection.prototype.onMyNodeChange_= function () {
...
 
 /**
+ * Detects when local client node is offline.  Responds by kicking the user
+ * out of the shard, which should update the interface and provide an
+ * opportunity to reconnect.
+ * @private
+ */
+NetSimConnection.prototype.onMyNodeLostConnection_ = function () {
+  this.disconnectFromShard();
+};
```
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimConnection.js]: Do we expect to do more her in the future? Could we just do setLostConnectionCallback(this.disconnectFromShard)?

Comment (islemaster) [inline apps/src/netsim/NetSimConnection.js]: Yes, I would like to do more here in the future; but that's the future.  Good call on the cleaner version for this iteration.

File: apps/src/netsim/NetSimHeartbeat.js [hunk 4]
```diff
@@ -133,12 +141,30 @@ NetSimHeartbeat.prototype.setBeatInterval = function (intervalMs) {
...
+/**
  * Updates own row on regular interval, as long as something's making
  * it tick.
  */
 NetSimHeartbeat.prototype.tick = function () {
   if (Date.now() - this.time_ > this.intervalMs_) {
     this.time_ = Date.now();
-    this.update();
+    this.update(function (success) {
+      if (!success) {
```
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimHeartbeat.js]: In node land, callbacks usually take the form callback(err, result). Might make more sense to emulate that and send an error (or no error) instead of success.

File: apps/src/netsim/NetSimHeartbeat.js [hunk 5]
```diff
@@ -133,12 +141,30 @@ NetSimHeartbeat.prototype.setBeatInterval = function (intervalMs) {
 };
 
 /**
+ * Set a handler to call if this heartbeat is unable to update its remote
+ * storage representation.  Can be used to go into a recovery mode,
+ * acknowledge disconnect, and/or attempt an auto-reconnect.
+ * @param {function} onFailedHeartbeat
+ */
+NetSimHeartbeat.prototype.setFailureCallback = function (onFailedHeartbeat) {
+  this.onFailedHeartbeat_ = onFailedHeartbeat;
```
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimHeartbeat.js]: Should we insert a check that a failedHeartbeat_ callback hasnt already been added?

File: apps/src/netsim/NetSimLocalClientNode.js [hunk 6]
```diff
@@ -163,76 +175,43 @@ NetSimLocalClientNode.prototype.tick = function (clock) {
...
-            // Reconnect failed
-            onComplete(false);
-            return;
-          }
+/**
+ * If a client update fails, should attempt an automatic reconnect.
+ * @param {function} [onComplete]
+ */
+NetSimLocalClientNode.prototype.update = function (onComplete) {
+  onComplete = (onComplete !== undefined) ? onComplete : function () {};
```
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimLocalClientNode.js]: Could simplify as onComplete = onComplete || function () {};

When using this pattern, you need to be careful about falsey values, but since your input is a function, that shouldnt be an issue.

Comment (islemaster) [inline apps/src/netsim/NetSimLocalClientNode.js]: Hm... I thought JSHint was complaining about that pattern, but now it isn't.  I'll change my default-setters over.

**Comment by Bjvanminnen** [comment]
Looks good for the most part. I would change your callbacks to take an option error instead of a success bool, as I think that's more standard in the JS world.

File: apps/src/netsim/NetSimHeartbeat.js (hunk 4 above)
Comment (islemaster) [inline apps/src/netsim/NetSimHeartbeat.js]: Sure.  I wasn't familiar with that convention; I was sort of following what Geoffrey set up with the storage table wrapper methods so callbacks could just be passed all the way down, but I'd be willing to switch things over.

Do you have a reference for this convention?  Is there a certain type of object that gets passed to `err` (like an actual `Error`)?  If there's no error, do these callbacks expect `null` or `undefined`?

Comment (Bjvanminnen) [inline apps/src/netsim/NetSimHeartbeat.js]: null. This is def a convention in node. I'm somewhat less certain that it is in non-node JS.

Comment (Bjvanminnen) [inline apps/src/netsim/NetSimHeartbeat.js]: Here's a reference https://docs.nodejitsu.com/articles/errors/what-are-the-error-conventions

**Comment by islemaster** [comment]
@Bjvanminnen I have added a whole bunch of commits to this PR; they are all a nice gradual transition over to using Node-style callbacks in all of the NetSim core components.  Along the way I improved test coverage.  Would you glance through this?

File: apps/src/appsApi.js [hunk 7]
```diff
@@ -45,31 +56,53 @@ var ApiRequestHelper = function (baseUrl) {
...
-  }).done(function (result /*, text*/) {
-    callback(result);
-  }).fail(function (/*request, status, error*/) {
-    callback(failureValue);
+  }).done(function (data /*, textStatus, jqXHR*/) {
+    callback(null, data);
+  }).fail(function (jqXHR, textStatus, errorThrown) {
+    callback(
+        new Error('textStatus: ' + textStatus + '; errorThrown: ' + errorThrown),
+        failureValue);
```
Comment (Bjvanminnen) [inline apps/src/appsApi.js]: I'm not sure we need failureValue here. The way I've usually seen this done is that if we have an error, the result is either null or garbage (i.e. we cant depend on it being anything in particular). Instead we can have a callback that looks something like this:
```
foo.get(myLocalUrl, function (err, result) {
  if (err !== null) {
    // handle failure. if we need some sort of failure value it can live here
    return;
  }
  // handle success
}

Comment (islemaster) [inline apps/src/appsApi.js]: Ah; you're right, I should pull that.  I had left it in initially to make it easy to transition over to the node-style callbacks; most consumers could keep expecting the same result.  I think I've entirely removed that dependency now, I'll double-check and then tear this out if possible.

File: apps/src/netsim/NetSimConnection.js [hunk 8]
```diff
@@ -94,7 +94,10 @@ var NetSimConnection = module.exports = function (sentLog, receivedLog) {
   this.statusChanges = new ObservableEvent();
 
   // Bind to onBeforeUnload event to attempt graceful disconnect
-  window.addEventListener('beforeunload', this.onBeforeUnload_.bind(this));
+  // Null-guard so constructing this object is test-friendly.
+  if (window && window.addEventListener) {
+    window.addEventListener('beforeunload', this.onBeforeUnload_.bind(this));
+  }
```
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimConnection.js]: Is there an equivalent way we can do this in tests? Right now it means there's a divergence between the code that runs in our app, and the code that we test. May not be a big deal in this case, but wold be nice to avoid these divergences if possible.

File: apps/src/appsApi.js (hunk 7 above)
Comment (Bjvanminnen) [inline apps/src/appsApi.js]: Cool. In general, I think it's good to avoid depending on the result in the callback to be a particular value when we have a non-null error.

File: apps/src/netsim/NetSimConnection.js [hunk 9]
```diff
@@ -173,34 +176,24 @@ NetSimConnection.prototype.disconnectFromShard = function () {
...
+      this.myNode.setLostConnectionCallback(this.disconnectFromShard.bind(this));
       this.myNode.initializeSimulation(this.sentLog_, this.receivedLog_);
-      this.myNode.update(function () {
+      this.myNode.update(function (/*err, result*/) {
         this.shardChange.notifyObservers(this.shard_, this.myNode);
         this.statusChanges.notifyObservers();
       }.bind(this));
     } else {
+      logger.error(err.message);
       logger.error("Failed to create client node.");
```
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimConnection.js]: I would structure this:
```
if (err) { // or if (err !== null)
  logger.error()
  return;
}
// success code
```
Reduces nesting and means you can stop thinking about errors as you're past that first block.

File: apps/src/netsim/NetSimConnection.js (hunk 8 above)
Comment (islemaster) [inline apps/src/netsim/NetSimConnection.js]: I was adding an empty addEventListener method to window in tests for a bit, but linting complained about the window object being readonly.  I'll change this up so we have to pass window into NetSimConnection; that way we can construct a fake.

File: apps/src/netsim/NetSimEntity.js [hunk 10]
```diff
@@ -75,35 +66,39 @@ NetSimEntity.create = function (EntityType, shard, onComplete) {
...
   var entity = new EntityType(shard);
-  entity.getTable_().read(entityID, function (row) {
-    if (row) {
-      onComplete(new EntityType(shard, row));
+  entity.getTable_().read(entityID, function (err, row) {
+    if (err === null) {
+      onComplete(null, new EntityType(shard, row));
     } else {
-      onComplete(null);
+      onComplete(err, null);
```
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimEntity.js]: Could restructure this as:
```
var entityType;
if (!err) { // i think i like this better than err === null, but both are fine
  entityType = new EntityType();
}
onComplete(err, entityType);

File: apps/src/netsim/NetSimEntity.js [hunk 11]
```diff
@@ -75,35 +66,39 @@ NetSimEntity.create = function (EntityType, shard, onComplete) {
...
- * @param {function} [onComplete] - Optional success callback.
+ * @param {NodeStyleCallback} [onComplete] - Optional success callback.
  */
 NetSimEntity.prototype.update = function (onComplete) {
   onComplete = onComplete || function () {};
 
-  this.getTable_().update(this.entityID, this.buildRow_(), onComplete);
+  this.getTable_().update(this.entityID, this.buildRow_(), function (err, result) {
+    onComplete(err, result);
+  });
```
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimEntity.js]: is this change necessary?

File: apps/src/netsim/NetSimEntity.js [hunk 12]
```diff
@@ -75,35 +66,39 @@ NetSimEntity.create = function (EntityType, shard, onComplete) {
...
 };
 
 /** Remove entity from remote storage. */
 NetSimEntity.prototype.destroy = function (onComplete) {
   onComplete = onComplete || function () {};
 
-  this.getTable_().delete(this.entityID, onComplete);
+  this.getTable_().delete(this.entityID, function (err, result) {
+    onComplete(result);
+  });
```
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimEntity.js]: Does this callback not take an error?

File: apps/src/netsim/NetSimHeartbeat.js [hunk 13]
```diff
@@ -133,12 +150,34 @@ NetSimHeartbeat.prototype.setBeatInterval = function (intervalMs) {
...
+
+/**
  * Updates own row on regular interval, as long as something's making
  * it tick.
  */
 NetSimHeartbeat.prototype.tick = function () {
   if (Date.now() - this.time_ > this.intervalMs_) {
     this.time_ = Date.now();
-    this.update();
+    this.update(function (err/*, result*/) {
```
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimHeartbeat.js]: Not sure if there's any value in this commented out result. I'd either just leave it in there uncommented (and let minifier remove it) or just remove it entirely.

File: apps/src/netsim/NetSimHeartbeat.js [hunk 14]
```diff
@@ -133,12 +150,34 @@ NetSimHeartbeat.prototype.setBeatInterval = function (intervalMs) {
...
+      if (err !== null) {
+        // A failed heartbeat update may indicate that we've been disconnected
+        // or kicked from the shard.  We may want to take action.
+        if (this.onFailedHeartbeat_ !== undefined) {
+          this.onFailedHeartbeat_();
+        }
+      }
+    }.bind(this));
```
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimHeartbeat.js]: Could reduce nesting by having if (err === null) return, then your failure code.

An argument to be made in favor of your approach is that there's value in being consistent across all of our callbacks in always starting by checking for/handling the non-null error case.

File: apps/src/netsim/NetSimLocalClientNode.js [hunk 15]
```diff
@@ -289,7 +273,9 @@ NetSimLocalClientNode.prototype.disconnectRemote = function (onComplete) {
 
     self.myWire = null;
     // Trigger an immediate router update so its connection count is correct.
-    self.myRouter.update(onComplete);
+    self.myRouter.update(function (err, result) {
+      onComplete(result);
+    });
```
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimLocalClientNode.js]: This and the change above it seem to be a couple more places where our callback doesnt take an error. Might be good to have it use callback(err, result) just for consistency, even if we'll never send it an error.

File: apps/src/netsim/NetSimMessage.js [hunk 16]
```diff
@@ -69,8 +69,8 @@ NetSimMessage.send = function (shard, fromNodeID, toNodeID, payload, onComplete)
   entity.fromNodeID = fromNodeID;
   entity.toNodeID = toNodeID;
   entity.payload = payload;
-  entity.getTable_().create(entity.buildRow_(), function (row) {
-    onComplete(row !== undefined);
+  entity.getTable_().create(entity.buildRow_(), function (err) {
+    onComplete(err === null);
```
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimMessage.js]: another place where i'd use callback(err, result)

Comment (Bjvanminnen) [inline apps/src/netsim/NetSimMessage.js]: or maybe we are just sending an err or no err? If that's the case, I'd maybe do onComplete(err, null)

Comment (Bjvanminnen) [inline apps/src/netsim/NetSimMessage.js]: I see there are a few more of these, so I'm going to stop commenting on each individually :)

File: apps/src/netsim/NetSimTable.js [hunk 17]
```diff
@@ -68,51 +68,75 @@ var NetSimTable = module.exports = function (storageTable) {
...
+ * @param {!NodeStyleCallback} callback
+ */
 NetSimTable.prototype.read = function (id, callback) {
-  this.remoteTable_.read(id, function (data) {
-    callback(data);
-    if (data !== undefined) {
+  this.remoteTable_.read(id, function (err, data) {
+    callback(err, data);
+    if (err === null) {
       this.updateCacheRow_(id, data);
```
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimTable.js]: Not sure how much it ends up mattering here, but it feels a little bit weird to callback and then do something. Usually I would expect us to do our own processing, then send it to the callback as the last thing we do.

Comment (Bjvanminnen) [inline apps/src/netsim/NetSimTable.js]: This pattern is repeated a couple times in this file.

**Comment by Bjvanminnen** [comment]
Mostly nits. Biggest thing I'd say is that it looks like there were a few places where we were having a single arg callback. I'd be consistent and always use callback(err, result), even if we never use the err. If we only use the err, I'd do callback(err, null).

File: apps/src/netsim/NetSimHeartbeat.js (hunk 13 above)
Comment (islemaster) [inline apps/src/netsim/NetSimHeartbeat.js]: I'll remove it.  The linter complains if you leave it in.

File: apps/src/netsim/NetSimHeartbeat.js (hunk 14 above)
Comment (islemaster) [inline apps/src/netsim/NetSimHeartbeat.js]: Yeah, I sort of wanted to leave this consistent with the other handlers; it's just unusual here that we only act in the failure case.

Comment (Bjvanminnen) [inline apps/src/netsim/NetSimHeartbeat.js]: Makes sense. Leaving it sounds good.

File: apps/src/netsim/NetSimLocalClientNode.js (hunk 15 above)
Comment (islemaster) [inline apps/src/netsim/NetSimLocalClientNode.js]: I figured I'd get the rest of these one-off cases as I went in future updates; I just wanted to change the core stuff now, to establish a baseline.  I can finish changing the remaining callbacks now, if you think it's worth doing.

File: apps/src/netsim/NetSimTable.js (hunk 17 above)
Comment (islemaster) [inline apps/src/netsim/NetSimTable.js]: Specifically, this is the shim between the storage API and the rest of the application which does local caching and provides change notifications.  The work we're doing after the callback is the change notification out to the rest of the system; if it wasn't being done immediately here, those notifications would come in later when the change came down from remote anyway.  I wanted to simulate that behavior as closely as possible by letting the initiating operation complete before kicking off the change notifications to other parts of the system.

File: apps/src/netsim/NetSimLocalClientNode.js (hunk 15 above)
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimLocalClientNode.js]: Up to you. I think it would be good to get to a place where they are all consistent, but if you think it makes more sense to make that transition over time, that's fine by me.

**Comment by islemaster** [comment]
Okay.  I went ahead and changed all the rest of the callbacks.
